internal/pkg/agent/cmd/upgrade.go

```diff
@@ -53,20 +58,73 @@ func newUpgradeCommandWithArgs(_ []string, streams *cli.IOStreams) *cobra.Comman
 	cmd.Flags().String(flagPGPBytes, "", "PGP to use for package verification")
 	cmd.Flags().String(flagPGPBytesURI, "", "Path to a web location containing PGP to use for package verification")
 	cmd.Flags().String(flagPGPBytesPath, "", "Path to a file containing PGP to use for package verification")
+	cmd.Flags().BoolP(flagForce, "", false, "Advanced option to force an upgrade on a fleet managed agent")
+	err := cmd.Flags().MarkHidden(flagForce)
+	if err != nil {
+		fmt.Fprintf(streams.Err, "error while setting upgrade force flag attributes: %s", err.Error())
+		os.Exit(1)
+	}
 
 	return cmd
 }
 
+type upgradeInput struct {
+	streams   *cli.IOStreams
+	cmd       *cobra.Command
+	args      []string
+	c         client.Client
+	agentInfo info.Agent
+}
+
 func upgradeCmd(streams *cli.IOStreams, cmd *cobra.Command, args []string) error {
 	c := client.New()
-	return upgradeCmdWithClient(streams, cmd, args, c)
+	agentInfo, err := info.NewAgentInfoWithLog(cmd.Context(), "error", false)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve agent info while tring to upgrade the agent: %w", err)
+	}
+	input := &upgradeInput{
+		streams,
+		cmd,
+		args,
+		c,
+		agentInfo,
+	}
+	return upgradeCmdWithClient(input)
 }
 
-func upgradeCmdWithClient(streams *cli.IOStreams, cmd *cobra.Command, args []string, c client.Client) error {
-	version := args[0]
+func shouldUpgrade(cmd *cobra.Command, agentInfo info.Agent) (bool, error) {
+	if agentInfo.IsStandalone() {
+		return true, nil
+	}
+
+	if agentInfo.Unprivileged() {
+		return false, fmt.Errorf("upgrade command needs to be executed as root for fleet managed agents")
+	}
+
+	force, err := cmd.Flags().GetBool(flagForce)
+	if err != nil {
+		return false, fmt.Errorf("failed to retrieve command flag information while trying to upgrade the agent: %w", err)
+	}
+
+	if !force {
+		return false, unsupportedUpgrade
+	}
+
+	return true, nil
+}
+
+func upgradeCmdWithClient(input *upgradeInput) error {
+	cmd := input.cmd
+	c := input.c
+	version := input.args[0]
 	sourceURI, _ := cmd.Flags().GetString(flagSourceURI)
 
-	err := c.Connect(context.Background())
+	su, err := shouldUpgrade(cmd, input.agentInfo)
+	if !su {
+		return fmt.Errorf("aborting upgrade: %w", err)
+	}
```

Comment:
The error wasn't even checked here :sweat_smile:, we can change this to 
```suggestion
	force, err := cmd.Flags().GetBool(flagForce)
	if err != nil {
		return fmt.Errorf("failed to retrieve command flag information while trying to upgrade the agent: %w", err)
	}

	err := checkUpgradable(force, input.agentInfo)
	if err != nil {
		return fmt.Errorf("aborting upgrade: %w", err)
	}
```